Use executable path as plugin path

jvff opened the pull request:
Checklist for a PR:

* [ ] Describe the change in **`CHANGELOG.md`**. Only applicable if the change has any impact for a user.

pronebird commented:
Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[talpid-core/src/tunnel/mod.rs, line 220 at r1](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89#-L6vQ0tJ4kUoOTLYtrIr:-L6vQ0tJ4kUoOTLYtrIs:bv3m4bp) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6090dc06ac570bd2f31f030d4dfa12f93fc25a36/talpid-core/src/tunnel/mod.rs#L220)):*
> ```Rust
>             .parent()
>             .map(Path::to_path_buf)
>             .ok_or(ErrorKind::ExecutableHasNoParentDir.into())
> ```

I am a little bit confused about that error. How's it possible that executable has no parent directory? 

I can only see that happen if it's running from memory location but then `current_exe()` would fail anyway.

Thoughts?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[talpid-core/src/tunnel/mod.rs, line 215 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89#-L6vPd6a5WYQdn747MpS:-L6vPd6beafewK72g77s:berg65v) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6090dc06ac570bd2f31f030d4dfa12f93fc25a36/talpid-core/src/tunnel/mod.rs#L215)):*
> ```Rust
> 
>     fn get_executable_dir() -> Result<PathBuf> {
>         let exe_path = env::current_exe().chain_err(|| ErrorKind::ExecutablePathInaccessible)?;
> ```

Regarding error handling. I feel it's easy to have an explosion of errors that can virtually never happen. If you look at `get_resource_dir()` in `main.rs` do you think the approach we took there has a problem with it? If any error happens along the way we just default to the `.` path.

---

*[talpid-core/src/tunnel/mod.rs, line 220 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89#-L6vPnbtyG22IEUSL_-E:-L6vPnbtyG22IEUSL_-F:b1y08zu) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6090dc06ac570bd2f31f030d4dfa12f93fc25a36/talpid-core/src/tunnel/mod.rs#L220)):*
> ```Rust
>             .parent()
>             .map(Path::to_path_buf)
>             .ok_or(ErrorKind::ExecutableHasNoParentDir.into())
> ```

Strictly according to the types, this error can happen. But can it happen in practice? If the `exe_path` is a file (it should always be?) then it must have a parent, all files have a parent. The only path that return `None` on `parent()` should be `/` right?

Which means it should be fairly safe to either just `unwrap()` here, or to fallback to the fallback stated above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[talpid-core/src/tunnel/mod.rs, line 220 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89#-L6vQ0tJ4kUoOTLYtrIr:-L6vQg62z5stP97NPMEN:b-nwctjp) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6090dc06ac570bd2f31f030d4dfa12f93fc25a36/talpid-core/src/tunnel/mod.rs#L220)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

I am a little bit confused about that error. How's it possible that executable has no parent directory? 

I can only see that happen if it's running from memory location but then `current_exe()` would fail anyway.

Thoughts?
</blockquote></details>

What does "running from memory location" mean?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[talpid-core/src/tunnel/mod.rs, line 220 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89#-L6vPnbtyG22IEUSL_-E:-L6vRCyxLGAP2Z9Bzl6x:blby29j) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6090dc06ac570bd2f31f030d4dfa12f93fc25a36/talpid-core/src/tunnel/mod.rs#L220)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Strictly according to the types, this error can happen. But can it happen in practice? If the `exe_path` is a file (it should always be?) then it must have a parent, all files have a parent. The only path that return `None` on `parent()` should be `/` right?

Which means it should be fairly safe to either just `unwrap()` here, or to fallback to the fallback stated above.
</blockquote></details>

Which is why we use `pop()` instead of `parent()` over there, since it does not have a failure really, it just stays the same path if it does not have one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[talpid-core/src/tunnel/mod.rs, line 220 at r1](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89#-L6vQ0tJ4kUoOTLYtrIr:-L6vR_vV5im59qS04gLj:b-9c86m0) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6090dc06ac570bd2f31f030d4dfa12f93fc25a36/talpid-core/src/tunnel/mod.rs#L220)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

What does "running from memory location" mean?
</blockquote></details>

You can google "launching exe from a memory buffer"

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->

jvff commented:
The Travis CI build failed, but it was due to a compiler bug (https://github.com/rust-lang/rust/issues/48336). 

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[talpid-core/src/tunnel/mod.rs, line 215 at r1](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89#-L6vPd6a5WYQdn747MpS:-L6v_U4j0TE77m98LZrU:b-gr517w) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6090dc06ac570bd2f31f030d4dfa12f93fc25a36/talpid-core/src/tunnel/mod.rs#L215)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Regarding error handling. I feel it's easy to have an explosion of errors that can virtually never happen. If you look at `get_resource_dir()` in `main.rs` do you think the approach we took there has a problem with it? If any error happens along the way we just default to the `.` path.
</blockquote></details>

Agreed. I removed both errors and used the same approach as in `get_resource_dir()`. Thanks!

---

*[talpid-core/src/tunnel/mod.rs, line 220 at r1](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89#-L6vPnbtyG22IEUSL_-E:-L6v_pKj9VHvxMcom26g:b-i9e5ap) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6090dc06ac570bd2f31f030d4dfa12f93fc25a36/talpid-core/src/tunnel/mod.rs#L220)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Which is why we use `pop()` instead of `parent()` over there, since it does not have a failure really, it just stays the same path if it does not have one.
</blockquote></details>

Removed as well, since it should never happen.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89#-:-L6vGUBMcAGYrOmKLiSm:b-qneq2n)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[talpid-core/src/tunnel/mod.rs, line 220 at r1](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89#-L6vQ0tJ4kUoOTLYtrIr:-L6vcdDE6av_0-uDqVL8:b-3zubfm) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6090dc06ac570bd2f31f030d4dfa12f93fc25a36/talpid-core/src/tunnel/mod.rs#L220)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

You can google "launching exe from a memory buffer"
</blockquote></details>

This piece has been removed so I marked this thread as satisfied.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->

faern commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89#-:-L6vlHR9uwjbpguQMpx7:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jvff commented:
Should this change be described in the changelog? From my understanding this is mostly a developer visible feature, correct?

---

Review status: 1 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/89#-:-L6yDT6t8mVFB7I9B167:b4sp12r)*
<!-- Sent from Reviewable.io -->

faern commented:
It might also be noticed by power users. And it is a change in functionality. But I agree it could be classified as slightly borderline. I like the part you added, it's good.

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89#-:-L6zGkvaIwaf4sDqmxgw:b73ruzn)*
<!-- Sent from Reviewable.io -->

mvd-ows commented:
Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 1 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/89)*
<!-- Sent from Reviewable.io -->
